test/Dialect/Arc/dedup.mlir

```diff
@@ -0,0 +1,228 @@
+arc.define @SplitArgumentsC(%arg0: i4, %arg1: i4) -> i4 {
+  %0 = comb.and %arg1, %arg0 {SplitArguments} : i4
+  arc.output %0 : i4
+}
+
+// CHECK-LABEL: hw.module @SplitArguments
+hw.module @SplitArguments(%clock: i1, %x: i4, %y: i4) {
+  // CHECK-DAG: arc.state @SplitArgumentsA(%x, %x)
+  // CHECK-DAG: arc.state @SplitArgumentsA(%x, %y)
+  // CHECK-DAG: arc.state @SplitArgumentsA(%y, %x)
+  %0 = arc.state @SplitArgumentsA(%x) lat 0 : (i4) -> i4
+  %1 = arc.state @SplitArgumentsB(%x, %y) lat 0 : (i4, i4) -> i4
+  %2 = arc.state @SplitArgumentsC(%x, %y) lat 0 : (i4, i4) -> i4
+  // CHECK-NEXT: hw.output
+}
+// CHECK-NEXT: }
+
+
+
+// CHECK-LABEL: arc.define @WeirdSplitArgumentsA
+// CHECK-SAME:    %arg0: i4
+// CHECK-SAME:    %arg1: i4
+// CHECK-SAME:    %arg2: i4
+// CHECK-SAME:    %arg3: i4
+arc.define @WeirdSplitArgumentsA(%arg0: i4, %arg1: i4) -> i4 {
+  // CHECK-NEXT: comb.and %arg0, %arg0, %arg1, %arg2, %arg3, %arg3
+  // CHECK-NEXT: arc.output
+  %0 = comb.and %arg0, %arg0, %arg0, %arg0, %arg1, %arg1 {WeirdSplitArguments} : i4
+  arc.output %0 : i4
+}
+// CHECK-NEXT: }
+
+// CHECK-NOT: arc.define @WeirdSplitArgumentsB
+arc.define @WeirdSplitArgumentsB(%arg0: i4, %arg1: i4) -> i4 {
+  %0 = comb.and %arg0, %arg0, %arg0, %arg1, %arg1, %arg1 {WeirdSplitArguments} : i4
+  arc.output %0 : i4
+}
+
+// CHECK-NOT: arc.define @WeirdSplitArgumentsC
+arc.define @WeirdSplitArgumentsC(%arg0: i4, %arg1: i4) -> i4 {
+  %0 = comb.and %arg0, %arg0, %arg1, %arg0, %arg1, %arg1 {WeirdSplitArguments} : i4
+  arc.output %0 : i4
+}
+
+// CHECK-LABEL: hw.module @WeirdSplitArguments
+hw.module @WeirdSplitArguments(%clock: i1, %x: i4, %y: i4) {
+  // CHECK-DAG: arc.state @WeirdSplitArgumentsA(%x, %x, %x, %y)
+  // CHECK-DAG: arc.state @WeirdSplitArgumentsA(%x, %x, %y, %y)
+  // CHECK-DAG: arc.state @WeirdSplitArgumentsA(%x, %y, %x, %y)
+  %0 = arc.state @WeirdSplitArgumentsA(%x, %y) lat 0 : (i4, i4) -> i4
+  %1 = arc.state @WeirdSplitArgumentsB(%x, %y) lat 0 : (i4, i4) -> i4
+  %2 = arc.state @WeirdSplitArgumentsC(%x, %y) lat 0 : (i4, i4) -> i4
+  // CHECK-NEXT: hw.output
+}
+// CHECK-NEXT: }
+
+
+// CHECK-LABEL: arc.define @VariadicDiffsDontDedupA
+arc.define @VariadicDiffsDontDedupA(%arg0: i4, %arg1: i4, %arg2: i4) -> i4 {
+  %0 = comb.and %arg0, %arg1, %arg2 {VariadicDiffsDontDedup} : i4
+  arc.output %0 : i4
+}
+
+// CHECK-LABEL: arc.define @VariadicDiffsDontDedupB
+arc.define @VariadicDiffsDontDedupB(%arg0: i4, %arg1: i4) -> i4 {
+  %0 = comb.and %arg0, %arg1 {VariadicDiffsDontDedup} : i4
+  arc.output %0 : i4
+}
+
+// CHECK-LABEL: hw.module @VariadicDiffsDontDedup
+hw.module @VariadicDiffsDontDedup(%clock: i1, %x: i4, %y: i4, %z: i4) {
+  // CHECK-DAG: arc.state @VariadicDiffsDontDedupA(%x, %y, %z)
+  // CHECK-DAG: arc.state @VariadicDiffsDontDedupB(%x, %y)
+  %0 = arc.state @VariadicDiffsDontDedupA(%x, %y, %z) lat 0 : (i4, i4, i4) -> i4
+  %1 = arc.state @VariadicDiffsDontDedupB(%x, %y) lat 0 : (i4, i4) -> i4
+  // CHECK-NEXT: hw.output
+}
+// CHECK-NEXT: }
```

Comment:
These are some great tests!

It probably makes sense to add a few more for these cases:
* one with an `scf.if` or similar in an arc to see regions work properly
* one where (only) different attributes on ops hinder dedup
* one where (only) different types on ops hinder dedup

Reply:
Excellent suggestions! Added those test. And of course they uncovered an issue where dedup would just blatantly ignore regions... :innocent: